### ISSUE TITLE
asyncio not using ProactorEventLoop with reload enabled

### ISSUE THREAD

**Issue by lausfl** [zauberzeug/nicegui#3874]
### Description

Version information:
- OS: Windows 10.0.19045
- Python: 3.11.4
- NiceGUI: 2.3.0

I was unable to find any documentation or existing issues about this and I am not sure if I am doing something wrong here.

I want to create an application that uses NiceGUI and has a separate process running from which I can parse the output on stdout and do something with it.
The problem I encounter is that I am unable to start the process when the `reload` argument of `ui.run()` is set to `True` (a NotImplementedError exception is thrown by asyncio).
On further investigation I found that the asyncio event loop is set to `_WindowsSelectorEventLoop` when reload is enabled and to `ProactorEventLoop` when it's disabled.

I found the following issue, which seems to be related: https://github.com/encode/uvicorn/issues/1220

So I tried setting the event loop policy to try and force it to use a `ProactorEventLoop`, but to no avail.

Here's the MRE:
```python
import asyncio

asyncio.set_event_loop_policy(asyncio.WindowsProactorEventLoopPolicy())
from nicegui import ui, app

def on_packet(packet):
    print('packet received')

async def start_ping():
    print(type(asyncio.get_event_loop()).__name__)
    cmd = ['ping', '127.0.0.1']
    proc = await asyncio.create_subprocess_exec(*cmd, stdout=asyncio.subprocess.PIPE)
    async for line in proc.stdout:
        print(line)

ui.label('Hello NiceGUI!')

app.on_startup(start_ping)
ui.run(reload=True)
```

When I change the above code to `ui.run(reload=False)`, it works as intended.

**Comment by rodja** [comment]
Have you tried [using a main guard](https://github.com/zauberzeug/nicegui/wiki/FAQs#why-is-my-code-executed-twice)? The subprocess evaluates the main file again, but should not start another NiceGUI application.

**Comment by lausfl** [comment]
@rodja do you mean like this?
```python
import asyncio

asyncio.set_event_loop_policy(asyncio.WindowsProactorEventLoopPolicy())
from nicegui import ui, app

def on_packet(packet):
    print('packet received')

async def start_ping():
    print(type(asyncio.get_event_loop()).__name__)
    cmd = ['ping', '127.0.0.1']
    proc = await asyncio.create_subprocess_exec(*cmd, stdout=asyncio.subprocess.PIPE)
    async for line in proc.stdout:
        print(line)

if __name__ == '__mp_main__':
    ui.label('Hello NiceGUI!')
    app.on_startup(start_ping)

ui.run(reload=True)
```
I also tried putting `app.on_startup(start_ping)` inside the guard. Either way I get the same behaviour.

**Comment by rodja** [comment]
The `ui.run()` must be guarded:

```py
...
if __name__ in ('__mp_main__', '__main__'):
    ui.run()
```

**Comment by lausfl** [comment]
It's still using the _WindowsSelectorEventLoop with the guard. I also tried putting all three lines inside the guard, but nothing changes.
Is this only happening on my end or can you reproduce it?

```python
import asyncio

asyncio.set_event_loop_policy(asyncio.WindowsProactorEventLoopPolicy())
from nicegui import ui, app

def on_packet(packet):
    print('packet received')

async def start_ping():
    print(type(asyncio.get_event_loop()).__name__)
    cmd = ['ping', '127.0.0.1']
    proc = await asyncio.create_subprocess_exec(*cmd, stdout=asyncio.subprocess.PIPE)
    async for line in proc.stdout:
        print(line)

ui.label('Hello NiceGUI!')
app.on_startup(start_ping)

if __name__ in ('__mp_main__', '__main__'):
    ui.run()
```

**Comment by rodja** [comment]
We do not use Windows. Maybe someone from the communty?

**Comment by falkoschindler** [comment]
@lausfl Is the problem still persisting? Can anyone else reproduce it?

**Comment by evnchn** [comment]
It is basically a matter-of-fact that when uvicorn is run with `reload=True` (yes, I should remind you that the hot-reload functionality comes from uvicorn, not us), that reload=True ==> SelectorEventLoop, reload=False ==> ProactorEventLoop. 

> Force the loop policy to `SelectorEventLoop` on Windows with 3.8+
https://github.com/encode/uvicorn/issues/529#issuecomment-567862449
https://github.com/encode/uvicorn/pull/535/files

---

So, you can either: 

- Mess with Uvicorn to un-force the loop policy to `SelectorEventLoop`
- Accept the fact and use reload=False

---

Also: I can't seem to be able to use WinLoop for this. I am still investigating why.

**Comment by evnchn** [comment]
@lausfl Check this out. WinLoop to the rescue, and now we force uvicorn to not set the event loop in any way by passing loop="none". 

<img width="1280" alt="Image" src="https://github.com/user-attachments/assets/799d3fe7-c1f4-4beb-a00a-028525728482" />

Source code: 
```py
import asyncio
from nicegui import ui, app
from fastapi import FastAPI
import winloop

fapp = FastAPI()

@ui.page("/")
def main_page():
    ui.label(str(type(asyncio.get_event_loop())))

async def start_ping():
    print(str(type(asyncio.get_event_loop())))
    cmd = ['ping', '127.0.0.1']
    proc = await asyncio.create_subprocess_exec(*cmd, stdout=asyncio.subprocess.PIPE)
    async for line in proc.stdout:
        print(line)

app.on_startup(start_ping)

asyncio.set_event_loop(winloop.new_event_loop())
asyncio.set_event_loop_policy(winloop.EventLoopPolicy())

ui.run(loop="none", port=9000) # let us configure the loop ourselves
```

---

Regarding https://github.com/zauberzeug/nicegui/discussions/4474: I may have to update my code snippet to fully use WinLoop. Apparently just doing winloop.install() (which is `asyncio.set_event_loop_policy(winloop.EventLoopPolicy())` in disguise) may not be enough in fully setting everything to use WinLoop.

**Comment by lausfl** [comment]
@falkoschindler yes, I just tested it with the latest version of nicegui and it still occurs.

**Comment by lausfl** [comment]
@evnchn thanks a lot for the alternative solution. I am closing the issue, as my problem can be solved by using WinLoop.

**Comment by lausfl** [comment]
@evnchn WinLoop works as expected, thanks!

**Comment by yamanjain** [comment]
Can someone post how they made it work ?
I have the same issue, trying Playwright with Nicegui (Hot-reload on)
